### PR TITLE
Update to the standard function docs

### PR DESCRIPTION
**Pull request by TristonianJones**
Provide more useful documentation for the CEL types, functions, and macros

### CODE REVIEW

File: doc/langdef.md [hunk 1]
```diff
@@ -1203,1342 +1193,925 @@ names are like `Europe/Paris`, `CET`, or `US/Central`.
...
+
+**list** `type(list(dyn))` \- Type denotation
+
+**map** `type(map(dyn, dyn))` \- Type denotation
+
+**null\_type** `type(null)` \- Type denotation
+
+**string** `type(string)` \- Type denotation
+
+**Signatures:**
```
Comment (jnthntatum) [inline doc/langdef.md:2040]: is string(bool) -> string part of the standard?

File: doc/langdef.md [hunk 2]
```diff
@@ -1203,1342 +1193,925 @@ names are like `Europe/Paris`, `CET`, or `US/Central`.
...
+
+**Examples:**
+
+```
+1 != 2     // true
+"a" != "a" // false
+3.0 != 3.1 // true
+```
+
+**Less Than or Equal To (\<=)** \- Compares two values and returns `true` if the
```
Comment (jnthntatum) [inline doc/langdef.md:1523]: Should this include a note about cross-numeric type comparisons?

File: doc/langdef.md [hunk 3]
```diff
@@ -941,46 +930,46 @@ and type of arguments and the type of the result, as well as an opaque
...
-  </tr>
-  <tr>
-    <td>
-      (map(A, B)) -> int
-    </td>
-    <td>
-      map size
-    </td>
-  </tr>
+ <tr>
```
Comment (jnthntatum) [inline doc/langdef.md:938]: Did you mean to de-indent here?

File: doc/langdef.md [hunk 4]
```diff
@@ -530,19 +520,18 @@ Object    | `map` (with string keys, bi-convertible values)
 We define JSON mappings for much of the remainder of CEL data, but note that
 this data will not map back in to CEL as the same value:
 
-CEL Data                                               | JSON Data
------------------------------------------------------- | ---------
-`int`                                                  | Number if in interoperable range, otherwise decimal String.
-`uint`                                                 | Number if in interoperable range, otherwise decimal String.
+CEL Data | JSON Data
+------------------------------------------------------ | --------- `int` |
```
Comment (jnthntatum) [inline doc/langdef.md:524]: I think this table is broken now.

File: doc/langdef.md (hunk 1 above)
Comment (TristonianJones) [reply doc/langdef.md:2049]: Good question. Seems like it should be. It wasn't in the table, but I'll check the implementations.

File: doc/langdef.md (hunk 2 above)
Comment (TristonianJones) [reply doc/langdef.md:1532]: I'll add a note at the top of the Comparisons section about cross-numeric type comparisons at runtime. Good call.

**Comment by TristonianJones** [comment]
@jnthntatum PTAL

File: doc/langdef.md [hunk 5]
```diff
@@ -1208,1337 +1207,938 @@ Regular expressions follow the
...
+[1, 2, 3].filter(x, x > 1) // [2, 3]
+['cat', 'dog', 'bird', 'fish'].filter(pet, pet.size() == 3) // ['cat', 'dog']
+[{'a': 10, 'b': 5, 'c': 20}].map(m, m.filter(key, m[key] > 10)) // [['c']]
+```
+
+**map** \- Returns a list where each element is the result of applying the
+transform expression to the corresponding input list element or input map key.
+
+There are both three argument and four argument forms of the macro. The three
+argument forms transform all elements. The four argument forms transform only
```
Comment (jnthntatum) [inline doc/langdef.md:1311]: grammar / clarity nit: pluralization is really odd in this description.

Something like:

For both maps and lists, the macro has a three argument and four argument form:
- The three argument form transforms all elements.
- The four argument form transforms only elements which satisfy the predicate. This is intended to simplify combined filter and map operations.

File: doc/langdef.md [hunk 6]
```diff
@@ -1208,1337 +1207,938 @@ Regular expressions follow the
...
+
+```
+['hello', 'world'].size() // 2
+size(['first', 'second', 'third']) // 3
+```
+
+#### Map Operators
+
+**Map Indexing (\[\])** \- map indexing. For string keys, cost is proportional
+to the size of the map keys times the size of the index string
```
Comment (jnthntatum) [inline doc/langdef.md:1680]: Up to you: this is technically correct, but practically this is always a hash map so maybe worth suggesting expected O(1) in practice as with the `in` description?

File: doc/langdef.md [hunk 7]
```diff
@@ -1208,1337 +1207,938 @@ Regular expressions follow the
...
+
+**Examples:**
+
+```
+"hello world".contains("world") // true
+"foobar".contains("baz") // false
+```
+
+**endsWith** \- Tests whether the string operand ends with the specified suffix.
+Time complexity is proportional to the product of the sizes of the arguments.
```
Comment (jnthntatum) [inline doc/langdef.md:1755]: Is this correct? Are we assuming no random access on the haystack string?

**Review by jnthntatum — APPROVED** [review]
Some  optional suggestions for clarifications. otherwise LGTM

**Review by jnthntatum — APPROVED** [review]

**Comment by sergiitk** [comment]
Thanks for the useful documentation update, @TristonianJones! 
One remark: I still found the old table format incredibly convenient as a concise overview of what methods are a part of the standard library. Would you consider bringing back the table at the top of the "Standard Definitions" sections, above the new in-detail docs?

**Pull request merged**